tests/pubsub/testgossipsub.nim

```diff
@@ -769,3 +769,103 @@ suite "GossipSub":
     )
 
     await allFuturesThrowing(nodesFut.concat())
+
+  asyncTest "GossipsSub peers disconnections mechanics":
+    var runs = 10
+
+    let
+      nodes = generateNodes(runs, gossip = true, triggerSelf = true)
+      nodesFut = nodes.mapIt(it.switch.start())
+
+    await allFuturesThrowing(nodes.mapIt(it.start()))
+    await subscribeNodes(nodes)
+
+    var seen: Table[string, int]
+    var seenFut = newFuture[void]()
+    for i in 0..<nodes.len:
+      let dialer = nodes[i]
+      var handler: TopicHandler
+      closureScope:
+        var peerName = $dialer.peerInfo.peerId
+        handler = proc(topic: string, data: seq[byte]) {.async, gcsafe, closure.} =
+          if peerName notin seen:
+            seen[peerName] = 0
+          seen[peerName].inc
+          check topic == "foobar"
+          if not seenFut.finished() and seen.len >= runs:
+            seenFut.complete()
+
+      dialer.subscribe("foobar", handler)
+      await waitSub(nodes[0], dialer, "foobar")
+
+    # ensure peer stats are stored properly and kept properly
+    check:
+      GossipSub(nodes[0]).peerStats.len == runs - 1 # minus self
+
+    tryPublish await wait(nodes[0].publish("foobar",
+                                  toBytes("from node " &
+                                  $nodes[0].peerInfo.peerId)),
+                                  1.minutes), 1, 5.seconds
+
+    await wait(seenFut, 5.minutes)
+    check: seen.len >= runs
+    for k, v in seen.pairs:
+      check: v >= 1
+
+    for node in nodes:
+      var gossip = GossipSub(node)
+      check:
+        "foobar" in gossip.gossipsub
+        gossip.fanout.len == 0
+        gossip.mesh["foobar"].len > 0
+
+    debug "Removing some subscriptions"
+
+    for i in 0..<runs:
+      if i mod 3 != 0:
+        nodes[i].unsubscribeAll("foobar")
+
+    debug "Waiting 2 heartbeats"
+
+    # wait for 2 heartbeats
+    for _ in 0..1:
+      for i in 0..<runs:
+        if i mod 3 == 0:
+          let evnt = newAsyncEvent()
+          GossipSub(nodes[i]).heartbeatEvents &= evnt
+          await evnt.wait()
+
+    # ensure peer stats are stored properly and kept properly
+    check:
+      GossipSub(nodes[0]).peerStats.len == runs - 1 # minus self
+
+    debug "Adding again subscriptions"
+
+    proc handler(topic: string, data: seq[byte]) {.async, gcsafe.} =
+      check topic == "foobar"
+
+    for i in 0..<runs:
+      if i mod 3 != 0:
+        nodes[i].subscribe("foobar", handler)
+
+    debug "Waiting 2 heartbeats"
```

Comment:
nitpick, maybe trace with a module level `logScope`?

Reply:
ah , to be fair I might just go further and remove those, they are not useful anymore